frontend/src/stimulus/controllers/dynamic/work-packages/progress/preview-progress.controller.ts

```diff
@@ -0,0 +1,113 @@
+import { debounce } from 'lodash';
+import morphdom from 'morphdom';
+
+interface TurboBeforeFrameRenderEventDetail {
+  render:(currentElement:HTMLElement, newElement:HTMLElement) => void;
+}
+
+export default class PreviewProgressController extends Controller {
+  static targets = [
+    'form', 'progressInput',
+  ];
+
+  declare readonly progressInputTargets:HTMLInputElement[];
+  declare readonly formTarget:HTMLFormElement;
+
+  private debouncedPreview:(event:Event) => void;
+  private frameMorphRenderer:(event:CustomEvent<TurboBeforeFrameRenderEventDetail>) => void;
+
+  connect() {
+    this.debouncedPreview = debounce((event:Event) => { void this.preview(event); }, 500);
+    // TODO: Ideally morphing in this single controller should not be necessary.
+    // Turbo supports morphing, by adding the <turbo-frame refresh="morph"> attribute.
+    // However, it has a bug, and it doesn't morphs when reloading the frame via javascript.
+    // See https://github.com/hotwired/turbo/issues/1161 . Once the issue is solved, we can remove
+    // this code and just use <turbo-frame refresh="morph"> instead.
+    this.frameMorphRenderer = (event:CustomEvent<TurboBeforeFrameRenderEventDetail>) => {
+      event.detail.render = (currentElement:HTMLElement, newElement:HTMLElement) => {
+        morphdom(currentElement, newElement, { childrenOnly: true });
+      };
+    };
+
+    this.progressInputTargets.forEach((target) => {
+      target.addEventListener('input', this.debouncedPreview);
+    });
+
+    const turboFrame = this.formTarget.closest('turbo-frame') as HTMLFrameElement;
+    turboFrame.addEventListener('turbo:before-frame-render', this.frameMorphRenderer);
+  }
+
+  disconnect() {
+    this.progressInputTargets.forEach((target) => target.removeEventListener('input', this.debouncedPreview));
+    const turboFrame = this.formTarget.closest('turbo-frame') as HTMLFrameElement;
+    turboFrame.removeEventListener('turbo:before-frame-render', this.frameMorphRenderer);
+  }
+
+  async preview(event:Event) {
+    const field = event.target as HTMLInputElement;
+    const form = this.formTarget;
+    const formData = new FormData(form) as unknown as undefined;
+    const formParams = new URLSearchParams(formData);
+    const wpParams = [
+      ['work_package[remaining_hours]', formParams.get('work_package[remaining_hours]') || ''],
+      ['work_package[estimated_hours]', formParams.get('work_package[estimated_hours]') || ''],
+      ['work_package[status_id]', formParams.get('work_package[status_id]') || ''],
+      ['field', field.name ?? 'estimatedTime'],
+      ['work_package[remaining_hours_touched]', formParams.get('work_package[remaining_hours_touched]') || ''],
+      ['work_package[estimated_hours_touched]', formParams.get('work_package[estimated_hours_touched]') || ''],
+      ['work_package[status_id_touched]', formParams.get('work_package[status_id_touched]') || ''],
+    ];
+
+    const wpPath = this.ensureValidPathname(form.action);
```

Comment:
I haven't tried it out by I suspect this will fail if the application is run with a relative url root like `/custom`. In this case, the matching would not work.